### PR TITLE
libxcrypt: Remaining changes

### PR DESCRIPTION
**Pull request by mweinelt**
###### Description of changes

Updates that I didn't want to push into #181764 anymore. These should be uncontroversial updates to the libxcrypt package and manual updates.

And an activation script that notifies users if they need to update their password hashes in `/etc/shadow`.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/config/supported-hashes.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    except ValueError:
+        # descrypt
+        return True
+
+    return scheme not in good_schemes
+
+
+users = set()
+try:
+    with open("/etc/shadow", "r") as fd:
```
Comment (mweinelt) [inline nixos/modules/config/supported-hashes.py:29]: I should probably grab `getent shadow` to account for other lookups. :thinking:

File: nixos/modules/config/supported-hashes.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python3
```
Comment (oxalica) [inline nixos/modules/config/supported-hashes.py:1]: This script could be simpler with a `bash` script using regex. It can also avoid pulling `python` dependency to the activation script.

```bash
users=()
while IFS=: read -r user hash tail; do
    if [[ "$hash" != '!' && ! "$hash" =~ ^\$(y|gy|7|2b|2y|2a|6)\$ ]]; then
        users+=("$user")
    fi
done </etc/shadow
```

File: nixos/doc/manual/configuration/user-mgmt.chapter.md [hunk 3]
```diff
@@ -32,8 +32,7 @@ account will cease to exist. Also, imperative commands for managing users and
 groups, such as useradd, are no longer available. Passwords may still be
 assigned by setting the user\'s
 [hashedPassword](#opt-users.users._name_.hashedPassword) option. A
-hashed password can be generated using `mkpasswd -m
-  sha-512`.
+hashed password can be generated using `mkpasswd -m yescrypt`.
```
Comment (oxalica) [inline nixos/doc/manual/configuration/user-mgmt.chapter.md:35]: What's the rationale for this change? sha-512 `$6$` is still allowed as `good_schemes`.
Also note that `yescrypt` is the type when calling `mkpassword` without `-m`.

**Review by oxalica — CHANGES_REQUESTED** [review]

File: nixos/modules/config/supported-hashes.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+if not users:
+    sys.exit(0)
+
+print(
+    """
+    WARNING: The following users are relying on password hashes that will
+    be removed in NixOS 23.05. They should be renewed as soon as possible!""",
+    file=sys.stderr,
+)
```
Comment (SuperSandro2000) [inline nixos/modules/config/supported-hashes.py:55]: Can we abort the switching in a next deprecation step?

Comment (winterqt) [reply nixos/modules/config/supported-hashes.py:55]: I also thought this was a good idea, but realized it's an obvious bad idea when doing `nixos-rebuild boot` into e.g. 23.05 -- the system would just fail to boot with that message, which... I don't know if we want that. Maybe we do? Not sure.

File: nixos/doc/manual/configuration/user-mgmt.chapter.md (hunk 3 above)
Comment (mweinelt) [reply nixos/doc/manual/configuration/user-mgmt.chapter.md:35]: I'll reference https://fedoraproject.org/wiki/Changes/yescrypt_as_default_hashing_method_for_shadow, where the author of libxcrypt explains why sha256crypt & sha512crypt are not optimal.

I'm aware it is the default, the ordering comes from the order in https://github.com/besser82/libxcrypt/blob/v4.4.28/lib/hashes.conf#L41.

I'm fine with removing the `-m` parameter entirely, and I'm against sticking with sha512 explicitly, because documentation gets outdated far too quickly.

File: nixos/modules/config/supported-hashes.py (hunk 2 above)
Comment (mweinelt) [reply nixos/modules/config/supported-hashes.py:1]: Thanks!

File: nixos/modules/config/supported-hashes.py (hunk 4 above)
Comment (mweinelt) [reply nixos/modules/config/supported-hashes.py:55]: We can harden the validation for `hashedPassword` using `strMatching`, but we don't have an equally good solution for mutable passwords.

**Comment by mweinelt** [comment]
Converting to draft while untested and with open change requests. If someone wants to work on the warning for `hashedPassword` please speak up.

**Comment by mweinelt** [comment]
Reduced to the uncontroversial changes, moving documentation and deprecation warning changes into dedicated PR.
